backend/asmgen.ml

```diff
@@ -285,6 +285,26 @@ let compile_unit ~output_prefix ~asm_filename ~keep_asm ~obj_filename gen =
          ~exceptionally:(fun () ->
              if create_asm && not keep_asm then remove_file asm_filename);
        if should_emit () then begin
+         (* Compact before calling the assembler, because the assembler can use
+            a lot of memory, and this reduces the peak memory usage by freeing
+            most of the memory from this process before the assembler starts
+            using memory.
+
+            On a run where this frees up around 1.1GB of memory this takes
+            around 0.6s. We only take this time on large jobs where the number
+            of parallel jobs is likely to be constrained by total system
+            memory. *)
+         let _minor, _promoted, major_words = Gc.counters () in
+         (* Uses [major_words] because it doesn't require a heap traversal to
+            compute and for this workload a majority of major words are live at
+            this point. *)
+         if major_words > 500_000_000.0 /. 8.0 then begin
+           reset ();
+           Typemod.reset ();
+           Emitaux.reset ();
+           Reg.reset ();
+           Profile.record_call "assemble_compact" Gc.compact;
```

Comment:
The condition on `major_words` before calling `Gc.compact` makes sense.  Why not resetting all of the others unconditionally?  How did you determine the hardcoded threshold for the condition?  Does it make sense to add a compilation flag that overrides it?

Reply:
(As we were discussing the threshold offline, I was pointed
to the equivalent in `Asmlink`. I would be nice to find a way
to share the threshold and the comments between the two
modules.)

Reply:
Emitaux is a good place to put the shared code from asmgen and asmlink.
Adding flambda-backend specific flags shouldn't be too painful now.

Reply:
I refactored into Emitaux.